contracts/HoprWhitehat.sol

```diff
@@ -0,0 +1,122 @@
+    bytes32 private constant TOKENS_RECIPIENT_INTERFACE_HASH = keccak256("ERC777TokensRecipient");
+
+    event RequestedGimme(address indexed account, uint256 indexed entitledReward);
+    event Called777Hook(uint256 indexed amount);
+    event Called777HookForFunding(uint256 indexed amount);
+    event ReceivedXHopr(address indexed account, uint256 indexed amount);
+    event ReclaimedBoost(address indexed account, uint256 indexed tokenId);
+
+    // entry function to be called by users who can unlock their tokens (users who have rewards)
+    function gimmeToken() external {
+        require(myHoprStake.owner() == address(this), "HoprStake needs to transfer ownership");        
+        // check 1820 implementation
+        require(ERC1820_REGISTRY.getInterfaceImplementer(msg.sender, TOKENS_RECIPIENT_INTERFACE_HASH) == address(this), "Caller has to set this contract as ERC1820 interface");
+        lastCaller = msg.sender;
+
+        myHoprStake.sync(lastCaller); // updates the rewards inside the accounts mapping struct
+        // solhint-disable-next-line no-unused-vars
+        (uint256 actualLockedTokenAmount, uint256 virtualLockedTokenAmount, uint256 lastSyncTimestamp, uint256 cumulatedRewards, uint256 claimedRewards) = myHoprStake.accounts(lastCaller);
+        uint256 stakerEntitledReward = cumulatedRewards - claimedRewards;
+        emit RequestedGimme(lastCaller, stakerEntitledReward);
+
+        wxHopr.send(address(myHoprStake), stakerEntitledReward, '0x0');
+
+        // unlock xHOPR
+        myHoprStake.unlock(lastCaller);
+    }
+    
+    // ERC777 fallback (wxHOPR aka reward tokens)
+    function tokensReceived(
+        address operator,
+        address from,
+        address to,
+        uint256 amount,
+        bytes calldata userData,
+        bytes calldata operatorData
+    ) external {
+        require(msg.sender == address(wxHopr), "can only be called from wxHOPR");
+        if (from == address(myHoprStake)) {            
+            require(to == address(this), "must send ERC777 tokens to HoprWhitehat");
+            emit Called777Hook(amount);
+            myHoprStake.reclaimErc20Tokens(address(xHopr));
+        }
+        else {
+            emit Called777HookForFunding(amount);
+        }
+    }
+
+    // ERC677 fallback (xHOPR aka token that users stake and unlock)
+    function onTokenTransfer(
+        address _from,
+        uint256 _value,
+        bytes memory _data
+    ) external returns (bool) {
+        require(msg.sender == address(xHopr), "only accept xHopr");
+        emit ReceivedXHopr(lastCaller, _value);
+        for (uint c = 0; c < myHoprStake.redeemedNftIndex(lastCaller); c++) {
+            uint tokenId = myHoprStake.redeemedNft(lastCaller,c);
+            emit ReclaimedBoost(lastCaller, tokenId);
+            myHoprStake.reclaimErc721Tokens(lastCaller, tokenId);
+            myHoprBoost.safeTransferFrom(address(this), lastCaller, tokenId);
+        }
```

Comment:
All of this could also be moved into the 777 hook and we leave this function empty. Might be cleaner, but it doesn't really make a difference.

Reply:
We actually don't need to do this dance. after `_claim`, `unlock` will proceed with returning NFTs.